### DISCUSSION TITLE
Ability to push changes to url without re-rendering

### DISCUSSION THREAD

**Discussion by matthewwolfe** [vercel/next.js#18072] [Help] [answered]
I am working on a project that has a page which works similar to google maps. As a user zooms/moves the map around, the url should be replaced with the updated center latitude and longitude values.

The url looks something like this:
`/@30.7921211,-101.7344277,7.23z` Where the first value is the lat, second is the long

And as the map is moved, these values change. 

I've been trying to use the nextjs router `replace` to update the url, but every call to `router.replace` causes the application to re-render. I am able to prevent re-rendering by using `history.pushState`, but that breaks routing with the back button: ie. after `pushState` has been called, the nextjs router cannot correctly update and re-render a new page when the back button is pressed.

So either:
1. I use `router.replace` which causes tthe application to re-render
2. I use `history.pushState` to prevent re-rendering, but that breaks other aspects of navigation

Is there any way that I can use NextJS routing to basically say: I want to update the url to `X`, but don't attempt any rendering?

**Comment by jamesmosier** [comment]
> causes the application to re-render

Do you mean React internally renders again? Or do you mean the page refreshes/reloads and visually repaints?


Also, have you taken a look at [shallow routing](https://nextjs.org/docs/routing/shallow-routing)?

**Reply by matthewwolfe** [reply]
I have looked at shallow routing, but that has not solved the issue for me. React internally renders again, and parts of the application that should not be affected are re-rendering again. 

Here is a picture of the profile when using `Router.replace(href, as, { shallow: true })` when going performing the following:
original url: `/map?lat=50&lng=50` => `/map?lat=40&lng=40`:

<img width="1375" alt="Screen Shot 2020-10-21 at 11 18 06 AM" src="https://user-images.githubusercontent.com/2324850/96740910-4a232380-138f-11eb-9183-5dfb64941458.png">

I would not expect the entire application to be re-rendering here.

Alternatively, the profiler does not record any re-rendering when performing the same action using `window.history.replaceState`, which makes sense, but I would like to be able to achieve that with the nextjs router if possible.

This is what I am currently using to avoid any re-rendering:
```window.history.replaceState(window.history.state, "", href);```

**Reply by jamesmosier** [reply]
Unfortunately with `next/router` you won't be able to avoid React re-rendering. `next/router` uses the Context API and updates its internal state to reflect whatever `router.*` method you call.

It hard to say what could help in this situation without seeing the code, but if you want to update the url without the page being effected your only choice would be using the `window.history` API as you mentioned.

Sorry I'm not able to help more 🙏

**Reply by matthewwolfe** [reply]
That's what I figured, and the confirmation of that from you is comforting.

**Comment by matthewwolfe** [comment, marked as answer]
Leaving this comment here in case others have the same question in the future:

I ended up going with 

``` 
window.history.replaceState({ ...window.history.state, as: newUrl, url: newUrl }, '', newUrl);
``` 

for when I wanted to update the url without doing any re-rendering. I also check against `Router.pathname` and only use `history.replaceState` when remaining on the same page. Otherwise I use the nextjs Router.

Specifying `window.history.state` is mandatory because it maintains some nextjs state internals. If you pass `{}` instead of `window.history.state`, moving forward/back will potentially be broken. This was the hardest issue for me to fix.

**Reply by chas13** [reply]
Thank you!

**Reply by llezhava** [reply]
Thanks a lot!!!

**Reply by OscarVelandia** [reply]
Thanks!

**Reply by semy** [reply]
@matthewwolfe Nice thank you very much :)

Did you also find a way to navigate forwards and backwards with the browser navigation and thus trigger the re-rendering again?

Ty for short reply 👍

**Reply by matthewwolfe** [reply]
@semy I was able to use the `NextRouter` with `window.history.replaceState` and it rendered when I went back to previous pages. I think if you are using `window.history.push` you might have problems, but I haven't tried that. For me, as long as I passed `window.history.state` it maintained the correct NextJS internals and re-rendering worked

**Reply by ghost** [reply]
Thankssss !

**Reply by Azteker** [reply]
@matthewwolfe  I tried `window.history.replaceState`,  if there are query params, the url will be automatically changed back.

**Reply by matthewwolfe** [reply]
@Azteker Did you pass `window.history.state` as the first parameter? When I use this the url is always generated with the query params set as part of the string.

```
window.history.replaceState(window.history.state, '', '/some-url?test=1');
window.history.replaceState(window.history.state, '', '/some-url?test=2');
```

**Reply by Azteker** [reply]
> @Azteker Did you pass `window.history.state` as the first parameter? When I use this the url is always generated with the query params set as part of the string.
> 
> ```
> window.history.replaceState(window.history.state, '', '/some-url?test=1');
> window.history.replaceState(window.history.state, '', '/some-url?test=2');
> ```

@mattcarlotta  I called window.history.replaceState(null, null, location.pathname) in useEffect. after the change the url reverts right away. But if I run it in console after page loaded, it works properly.

**Reply by matthewwolfe** [reply]
@Azteker You need to pass `window.history.state` as the first parameter.

**Reply by ethanryan** [reply]
@matthewwolfe does the href need to be a string? I'm trying to use `window.history.replaceState(window.history.state, "", href)` with a URL object instead of an href string and can't get this to work.

**Reply by matthewwolfe** [reply]
@ethanryan I have only tried it with a string

**Reply by MikeWillis** [reply]
@matthewwolfe I'm using your suggestion to handle pagination on the product list of an ecommerce category page, so the URL changes without re-rendering the parent page. Works great and it's lightning fast due to only re-rendering what needs to be re-rendered (the product list). However, for some reason when I click the browser's back button, I get a full refresh of the entire app. Any idea how to avoid that? I'm continuing to tinker but I'm pretty unfamiliar with window.history so it's a lot of trial and error so far.

Edit: might be worth nothing that, for whatever reason, my tests show that this:

`window.history.replaceState({ ...window.history.state, as: newUrl, url: newUrl }, '', newUrl);`

and this:

`window.history.replaceState({}, '', newUrl);`

behave the same when browser back is clicked - both result in a full page refresh

**Reply by PCPbiscuit** [reply]
Thank you!!!

**Reply by nicholashamilton** [reply]
Just what I was looking for.. thanks

**Reply by mattcasey** [reply]
Still necessary, thanks!

**Reply by andretti1977** [reply]
that's the solution i've been looking for for two days!!! THANK YOU!!! I can't understand why this solution isn't explicitly stated in the docs since it's a very questioned topic but this is the only working solution

**Reply by ryparker** [reply]
This works great to prevent the full page re-render. Thank you! 

Has anyone found a way to trigger a re-render in another component whenever the URL params change using @matthewwolfe's solution?

For example, say I have a `<Toggle/>` component. When the `<Toggle/>` is clicked it will update the URL query using the `window.history.replaceState(...)` solution. But when the URL query updates I want my `<Results/>` component to update. Only the `<Toggle/>` and `<Results/>` components should re-render. Any ideas?

I've tested using a `useEffect` in my `<Results/>` component  that depends on `window.history.state`/`window.location.pathname` but no dice. I've also tried depending on next/router, specifically `router.query`.

**Reply by janghanp** [reply]
@ryparker 

This might help? -> https://developer.mozilla.org/en-US/docs/Web/API/Window/popstate_event

**Reply by martinharyanto** [reply]
hi @matthewwolfe can you explain what to provide as `newUrl`? should it be the whole url with the new query params? or just the query params? Thanks in advance 🙏

**Reply by matthewwolfe** [reply]
@martinharyanto whole, fully built, url

**Reply by vissaf** [reply]
> hi @matthewwolfe can you explain what to provide as `newUrl`? should it be the whole url with the new query params? or just the query params? Thanks in advance 🙏

If you're using Next.js, you can do something like this: 
```
const newUrl = `${
      router.asPath.includes("?")
        ? router.pathname + `?existingQuery=${router.query.existingQuery}` + `&newQuery=${formattedDate}`
        : router.pathname + `?newQuery=${formattedDate}`
    }`;
```

Hope this helps! 🙂

**Reply by bombillazo** [reply]
This was a Godsend, thank you all!!!!

**Reply by earthlingdavey** [reply]
> whole, fully built, url

Hey, path (and query) works ok for me, but I discovered a bug with using the full url (inc. protocol and host). 

To reproduce
1. Navigate one or more times with `window.history.replaceState`
1. Click a standard `next/link` - it works.
1. Click back, it doesn't work as expected. 
    The url reverts to `https://example.com/template` where it should be `https://example.com/template/slug1/slug2`

**Reply by kirandash** [reply]
The following works for me: `window.history.replaceState(window.history.state, '', newUrl);`

Hence I am curious why we need `as` and `url` in: `window.history.replaceState({ ...window.history.state, as: newUrl, url: newUrl }, '', newUrl);`

I don't see `as` and `url` prop in the `window.history.state` when I log it.

**Reply by JimCarnicelli** [reply]
Wow! I can vouch for the need for the spread operation and the "as" and "url" properties. You will not have a good time if you express it like this:

`window.history.replaceState(window.history.state, '', newUrl);`

I found when I would hit the back button on my browser to return to the page that did all the URL rewrites, the URL would be there for a moment. Then it would clear out to the original state with no query string (as I had altered). Very frustrating.

Good job. Thank you for sharing.

**Reply by mansigeek** [reply]
It's working..

**Reply by JCs3lma** [reply]
I am using `window.history.pushState` or `window.history.replaceState`, but when I try to access the parameters with `useParams`, it doesn't provide the updated parameters. I have opened a discussion about this; you can find it here: https://github.com/vercel/next.js/discussions/70779

Any workaround? I like to use `window.history.pushState` as it is more faster the `router.push` or router.replace

**Reply by earthlingdavey** [reply]
Hey @JCs3lma 👋 I found a novel workaround... `npm uninstall next`

**Reply by topolanekmartin** [reply]
Okay. Why did I come to this solution with version 15.3? 🤔 I don't like it at all, but it works.

```typescript
timer.current = setTimeout(() => {
    window.history.state.__PRIVATE_NEXTJS_INTERNALS_TREE[1][
	    'children'
    ][1]['children'][0][1] = newPathnameUrl
    
    window.history.replaceState(
	    window.history.state,
	    '',
	    newAbsoluteUrl,
    )
}, 500)
```

**Reply by GrahamQuan** [reply]
i learn this api from their `ai-chatbot` app

**Reply by sa-codymai** [reply]
For everyone that come later and found this no longer work.
Nextjs (tyranny!! 💢 ) has override the default `window.history.pushState`, `window.history.replaceState` in order to enforce re-rendering of server components when query string changes
Put following in your `src/app/index.ts` and use them instead
```ts
export const windowHistoryPushState = window.history.pushState;
export const windowHistoryReplaceState = window.history.replaceState;
```

**Comment by bamdadfr** [comment]
Having the same re-renders on `RouterContext` change, I started with the suggestion from @matthewwolfe and noticed `url` and `as` properties should have different values.

I ended up creating the [next-replace-url](https://github.com/bamdadsabbagh/next-replace-url) npm package to support this functionnality.

Browser action like back button works for me, no full page refresh.

**Reply by akash-mahmud** [reply]
Hi, 
I have checked his package. I have an issue. How can I send my full query object using this package? Because in this package it seems that it will just accept query parameres in key-value pairs. 
But how can I pass an object of multiple query parameters here like router.push?
Any help will be appreciated.

**Reply by bamdadfr** [reply]
Hi @akash-mahmud,
Thanks for looking into the package and for your feedback.
Your request is relevant, feel free to open a PR to support this functionality.

**Comment by 72L** [comment]
I am using:
```
window.history.replaceState({ ...window.history.state, as: newUrl, url: newUrl }, '', newUrl);
```
as suggested but it looks like the re-rendering is still happening. 

using app router, next version 13.4.10

**Reply by HelaGone** [reply]
I'm also having this issue with 14.0.4. the funny thing is that it only happens in production environment and not in localhost tho.

**Reply by kcrwfrd** [reply]
Next.js app router monkey patches `pushState` and `replaceState` specifically to trigger a re-render so that `usePathname()` and `useSearchParams()` will return the correct, updated URL when it changes - https://github.com/vercel/next.js/blob/5969a097b44eae1dd314e8f84d13733edf73af3e/packages/next/src/client/components/app-router.tsx#L476-L520

I don't think it will be possible to avoid re-render because of this, but you can accomplish at least accomplish the equivalent of a "shallow" transition by using `window.history.pushState` or `window.history.replaceState`.

**Comment by ajstokar** [comment]
I tried this but the back button isn't working for me.

**Reply by earthlingdavey** [reply]
I think you'll need pushState instead of replaceState.

https://stackoverflow.com/questions/17507091/replacestate-vs-pushstate

> replaceState() will change the URL in the browser (ie. pressing the back button won't take you back)
>
> pushState() will change the URL, and keep the old one in the browser history (ie. pressing the back button will take you back)

I would also advise thorough testing as push/replace state does not keep the nextjs router in sync. i.e. running `pushState` with a new path, then `router.asPath`, you will get incorrect path.

**Comment by federicoWax** [comment]
Neceisto hacer esto pero con link en nect 14, la prop shallow de Link no funciona para rsc

**Comment by ronny1020** [comment]
Is there any update after the next 14 with app router?

**Reply by antofa** [reply]
same, not working for next 14

**Comment by tvkiranm** [comment]
I have one issue with pages router when i click on the redirection then my page is refresh Why it behave like this in next js  
When I set the query param in the URL

**Comment by espensorby** [comment]
Has the issue with Next 14 and router.push/router.replace no longer accepting "{ shallow: true } still not been addressed?

**Comment by jeromevvb** [comment]
I am using Next 14.2.5. 
In my case, the user creates a new post by going to `/posts/new`
In the background, I created the post and changed `/new` with the ID of the created post
The following code works like a charm.

```
window.history.pushState({}, '', `/posts/${id}`)
```

**Reply by ekebobrown** [reply]
This worked for my use case where I needed to load a document server an ODM source onto a different component (preview) when a list item is clicked from a different component, both client components and imported into another component (server component). The idea being that when the URL changes the preview component will fetch and re-render using a parameter in the URL. 
I prefer the history's pushState for some reason being that when the user click the back button the last item previewed will be shown.

**Comment by David05500** [comment]
Any input on this from the official team? 

As doing a workaround like this seems like a bad practise, considering this functionality was working fine in page router


```
 const createQueryString = useCallback(
    (updates: { [key: string]: string }) => {
      const params = new URLSearchParams(searchParams);
      Object.entries(updates).forEach(([key, value]) => {
        params.set(key, value);
      });
      return params.toString();
    },
    [searchParams],
  );

  function prevPage() {
    const newPage = localPage > 1 ? localPage - 1 : 1;
    const newUrl = `${pathname}?${createQueryString({ page: newPage.toString(), limit: localLimit.toString() })}`;
    window.history.replaceState(
      { ...window.history.state, as: newUrl, url: newUrl },
      "",
      newUrl,
    );
    setLocalPage(newPage);
  }

```

**Reply by earthlingdavey** [reply]
Hi @David05500 ... it's been a while since I've read this issue, but I think it did/does affect page router also.

**Comment by kaurkaelep** [comment]
A replacement for { shallow: true }

it is really important that you tie your state to either usePathname() or/and useSearchParams().

```
const pathname = usePathname(),
 id = pathname.split('/').pop()
```

Whenever you call either pushState or replaceState, the pathname will be updated. 
`window.history.replaceState(null, '', newUrl)`

**Reply by ptvty** [reply]
Exactly. `usePathname` and `useSearchParams` work but `useParams` doesn't, which I don't see any valid reason why. As all three hooks are client-side hooks. I opened https://github.com/vercel/next.js/discussions/73656 for this.
